File: java/com/countrygamer/countrygamercore/common/handler/packet/PacketTeleport.java
package com.countrygamer.countrygamercore.common.handler.packet;

import com.countrygamer.core.Base.common.packet.AbstractPacket;
import com.countrygamer.countrygamercore.lib.CoreUtil;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import net.minecraft.entity.player.EntityPlayer;

public class PacketTeleport extends AbstractPacket {

	private int dimId = 0;
	private double[] coords = new double[3];
	private boolean fallDamage, particles;

	public PacketTeleport() {

	}

	public PacketTeleport(int dim, double[] coords) {
		this(dim, coords, false, false);
	}
	public PacketTeleport(int dim, double[] coords, boolean fall, boolean particles) {
		this.dimId = dim;
		this.coords = coords;
		this.fallDamage = fall;
		this.particles = particles;
	}

	@Override
	public void encodeInto(ChannelHandlerContext ctx, ByteBuf buffer) {
		buffer.writeInt(this.dimId);
		buffer.writeDouble(coords[0]);
		buffer.writeDouble(coords[1]);
		buffer.writeDouble(coords[2]);
		buffer.writeBoolean(this.particles);
		
	}

	@Override
	public void decodeInto(ChannelHandlerContext ctx, ByteBuf buffer) {
		this.dimId = buffer.readInt();
		this.coords[0] = buffer.readDouble();
		this.coords[1] = buffer.readDouble();
		this.coords[2] = buffer.readDouble();
		this.particles = buffer.readBoolean();
		
	}

	@Override
	public void handleClientSide(EntityPlayer player) {
		//System.out.println("Client PacketTeleport Recieved");
	}

	@Override
	public void handleServerSide(EntityPlayer player) {
		//System.out.println("Server PacketTeleport Recieved");
		CoreUtil.teleportPlayerToDimension(player, this.dimId);
		CoreUtil.teleportPlayer(player, coords[0], coords[1], coords[2], fallDamage, particles);
	}

}

File: src/main/scala/com/countrygamer/cgo/common/network/PacketTeleport.scala
package com.countrygamer.cgo.common.network

import com.countrygamer.cgo.common.lib.util.UtilVector
import com.countrygamer.cgo.wrapper.common.network.AbstractPacket
import io.netty.buffer.ByteBuf
import net.minecraft.entity.player.EntityPlayer

/**
 *
 *
 * @author CountryGamer
 */
class PacketTeleport(var dimId: Int, var coords: Array[Double], var fall: Boolean,
		var particles: Boolean) extends AbstractPacket {

	// Default Constructor

	// End Constructor

	// Other Constructors
	def this() {
		this(0, null, false, false)
	}

	// End Constructors
	override def writeTo(buffer: ByteBuf): Unit = {
		buffer.writeInt(this.dimId)
		buffer.writeDouble(coords(0))
		buffer.writeDouble(coords(1))
		buffer.writeDouble(coords(2))
		buffer.writeBoolean(this.fall)
		buffer.writeBoolean(this.particles)

	}

	override def readFrom(buffer: ByteBuf): Unit = {
		this.dimId = buffer.readInt()
		this.coords(0) = buffer.readDouble()
		this.coords(1) = buffer.readDouble()
		this.coords(2) = buffer.readDouble()
		this.fall = buffer.readBoolean()
		this.particles = buffer.readBoolean()

	}

	override def handleOnClient(player: EntityPlayer): Unit = {}

	override def handleOnServer(player: EntityPlayer): Unit = {
		UtilVector.teleportPlayerToDimension(player, this.dimId)
		UtilVector.teleportPlayer(player, coords(0), coords(1), coords(2), fall, particles)

	}

}
